--- a/drivers/misc/mediatek/thermal/mt8167/mtk_cooler_fps.c	2019-02-23 12:11:07.563369956 +0100
+++ b/drivers/misc/mediatek/thermal/mt8167/mtk_cooler_fps.c	2019-03-05 17:27:58.290658434 +0100
@@ -143,7 +143,10 @@
 bool  __attribute__ ((weak))
 mtk_get_gpu_loading(unsigned int *pLoading)
 {
-	pr_err("E_WF: %s doesn't exist\n", __func__);
+	/*
+	 * Arcadyan modified to avoid displaying too much message.
+	 */
+	//pr_err("E_WF: %s doesn't exist\n", __func__);
 	return 0;
 }
 
